### DISCUSSION TITLE
how to use api.Compiler().ConstantValue() ?

### DISCUSSION THREAD

**Discussion by YurunChen** [Consensys/gnark#568] [Q&A]
I tried to put frontend. Variable to *big. int, and then convert the data to fr. Element, I found api. Compiler(). ConstentValue() did not successfully convert my value to *big. Int。 I'm confused about the process and want to ask how this is going to be converted. Also, if you can, please check the Usage of  fr.Element. 
```
type Circuit struct {
	Mx   [][]frontend.Variable `gnark:"public"`
	R    frontend.Variable
	chal []frontend.Variable `gnark:"public"`
	tag  []frontend.Variable `gnark:"public"`
}

func (circuit *Circuit) Define(api frontend.API) error {

	Mx := make([][]*fr.Element, len(circuit.Mx))
	for i := 0; i < len(circuit.Mx); i++ {
		Mx[i] = make([]*fr.Element, len(circuit.Mx[i]))
		for j := 0; j < len(circuit.Mx[i]); j++ {
			// TODO why the v is nil? how to solve it?
			v, result := api.Compiler().ConstantValue(circuit.Mx[i][j])
			api.Println(v, result)
			Mx[i][j].SetBigInt(v)
		}
	}

	x := fr.Element{1}
	sigma := &x
	sigma.SetOne()
	PkSum := make([]*fr.Element, len(circuit.Mx))
	for i := range PkSum {
		PkSum[i].SetUint64(0)
	}
	tag := make([]*fr.Element, len(circuit.tag))
	for i, v := range circuit.tag {
		t, _ := api.Compiler().ConstantValue(v)
		tag[i].SetBigInt(t)
	}
	chal := make([]*fr.Element, len(circuit.chal))
	for i, v := range circuit.chal {
		t, _ := api.Compiler().ConstantValue(v)
		chal[i].SetBigInt(t)
	}
	for i, _ := range circuit.chal {
		t := tag[i].Mul(tag[i], chal[i])
		sigma = sigma.Mul(sigma, t)
		for j := 0; j < len(circuit.Mx); j++ {
			v := chal[i].Mul(chal[i], Mx[i][j])
			PkSum[i].Add(PkSum[i], v)
		}
	}

	return nil
}
```

**Comment by gbotrel** [comment]
```
Mx[i] = make([]*fr.Element, len(circuit.Mx[i]))
		for j := 0; j < len(circuit.Mx[i]); j++ {
			// TODO why the v is nil? how to solve it?
			v, result := api.Compiler().ConstantValue(circuit.Mx[i][j])
```
--> that's a Go question rather than specific to this lib; you initalize a slice of `[]*fr.Element` ; the default value is nil for pointers. Not sure what you're trying to do here; if you need to manipulate constants they don't need to be `frontend.Variable` they can be int, fr.Element, big.Int ,...

**Reply by YurunChen** [reply]
I want to use frontend. Variable does some arithmetic operations on finite fields like G1, G2, GT, Zr, but I don't know how to use these interfaces in the circuit. You know, when I input values from the outside to the circuit, their types change to frontend. variables, so I don't know how to manipulate them.

**Reply by hussein-aitlahcen** [reply]
Arithmetic operations in the circuit are done via the `api` interface, like `api.Add(x, y)` etc...
Similarly, control flow must also be done using the api if it's based on the input (somehow "dynamic"), `api.Select(condition, trueBranch, falseBranch)`
